### ISSUE TITLE
Add denoiser

### ISSUE THREAD

**Issue by gkjohnson** [gkjohnson/three-gpu-pathtracer#85]
https://github.com/BrutPitt/glslSmartDeNoise

**Comment by DennisSmolek** [comment]
The other day I was doing something trying to learn more about simulating clouds and I saw a promotion/release from Nvidia about their raytracing and it showed how they used denoise to reduce the sampling by at least half for even huge complex stuff. 

And that form most workflows, it was like the last node in the pipeline for super smooth results. 

I love your raytracer but I'm on a potato.. I can get to about 1000 passes but it still looks grainy... So I had shelved being able to use it personally any time soon.

I tried to see if there was a standard few sets of algorithms like there is for GENERATING noise, and it seems like theres some old school fundamentals but theres also a lot of crazy innovative things coming out all the time.. I tried limiting to things published within the last 3 years...

Of course found a big 0 for the WebGL space, but Nvidia and others are pushing each other. 

This paper came up a few times in discussions: https://cs.dartmouth.edu/wjarosz/publications/mara17towards.pdf

But when my eyes hit the math on page two:
<img width="825" alt="Screen Shot 2022-04-09 at 06 31 04" src="https://user-images.githubusercontent.com/1397052/162571045-4c19072f-9bfe-4e47-9745-f9e211066ee9.png">

I figured it reach out first haha..

Right before posting I found: https://alain.xyz/blog/ray-tracing-denoising

Which I didn't want to dive into if you had a solution in mind or strong opinions or something... 

After our tweets I'm going to go over that article, I pulled some others too. 

Dont search "Denoise" on shader tree.. .It repeatedly crashed my mac 100%, not just the browser..

There are 20-30 there too, that hopefully are implementations I can try

**Comment by gkjohnson** [comment]
> I love your raytracer but I'm on a potato.. I can get to about 1000 passes but it still looks grainy...

Ha some day I'd like this to work well even on a potato but it's still early. There's still a lot of work to do to improve the speed of convergence including multiple importance sampling, direct light sampling, and denoising. But I'd like help getting all of it going so I appreciate you taking a look into this a bit!

> Which I didn't want to dive into if you had a solution in mind or strong opinions or something...

I don't have a particular solution in mind. I think ideally a simple solution would come first and not require things like roughness / normal / depth texture output to aid in the noise reduction but I'm not sure how feasible that would be. I'm unfamiliar with the techniques used for this. That alain blog post looks pretty good. No rush on any of this, of course! Just getting an idea for what's possible is an awesome start.

In terms of API I was imagining something like a `DenoiseMaterial` that would be run just before rendering to the canvas. Effectively replacing the use of `MeshBasicMaterial` with the final full screen quad pass here:

https://github.com/gkjohnson/three-gpu-pathtracer/blob/99fc07a0ed53d8db3f50f857a562dcebbfeee5d2/example/materialBall.js#L90-L92

```js
const fsQuad = new FullScreenQuad( new DenoiseMaterial( {
  map: ptRenderer.target.texture,
  // denoise parameters
} ) );

// ...

fsQuad.render( renderer );
```

**Comment by DennisSmolek** [comment]
There

**Comment by DennisSmolek** [comment]
There are a few complex though well written out approaches/solutions as well such as:

["Caffeine" by Rye Terrell](https://wwwtyro.net/2018/02/25/caffeine.html)
[Demo](https://wwwtyro.github.io/caffeine/)

Where he uses [some kind] of multisample approach..

You may already be doing something similar, This is where I have to admit I hadn't dived through your full flow yet..


[Heres an older version of the Nvidia AI approach too](https://blogs.nvidia.com/blog/2017/12/07/nvidia-optix-ai-denoiser/)

**Comment by thomasaull** [comment]
Just to throw it in: Maybe it's possible to use the Intel Open Image Denoiser via wasm? https://www.openimagedenoise.org/

**Comment by DennisSmolek** [comment]
> Just to throw it in: Maybe it's possible to use the Intel Open Image Denoiser via wasm? https://www.openimagedenoise.org/

I'm not sure if jumping into WASM first is a good idea, but if one exists why not!
===
I've been meaning to break out the denoiser from the[ LGL Raytracer](https://lgltracer.com/)
Which is pretty legit and fast: https://lgltracer.com/cornellBox/index.html

Paul recently showed off a demo and actually made connectors for it in react
Demo: https://codesandbox.io/s/lgl-raytracer-rnuve
Library: https://github.com/pmndrs/react-three-lgl

Oddly too though the creator of the LGL has decided to nolonger update it on NPM: https://www.npmjs.com/package/lgl-tracer

But I think it wouldnt be too hard to pull the denoise out and try it out at least

**Comment by gkjohnson** [comment]
I agree I'd like to skip using WASM for now but there's probably a lot to learn from the open image denoise project. The source for it is here:

https://github.com/OpenImageDenoise/oidn

And it looks like the [GLSL-PathTracer](https://github.com/knightcrawler25/GLSL-PathTracer) project uses it, as well.

**Comment by gkjohnson** [comment]
Here's a few documents on SVGF which might be promising for this use case:

https://teamwisp.github.io/research/svfg.html

https://cg.ivd.kit.edu/publications/2017/svgf/svgf_preprint.pdf

**Comment by DennisSmolek** [comment]
The "Current" leading methods according to resources like that page have you doing crazy stuff with machine learning etc..

There's lots we can do before that, my current focus is this paper: https://webpages.tuni.fi/foi/papers/Koskela-TOG-2019-Blockwise_Multi_Order_Feature_Regression_for_Real_Time_Path_Tracing_Reconstruction.pdf
Which calls its method Blockwise Multi-Order Feature Regression (BMFR)
And claims a 1.8X performance gain over the previous generation, notably SVGF

Which is a common method and the one used by LPL

SVGF paper: https://cg.ivd.kit.edu/publications/2017/svgf/svgf_preprint.pdf

Basically we mix lighting and raster by splitting things.


This was a WIP for the existing three raytracing-renderer

https://github.com/hoverinc/ray-tracing-renderer/issues/2

And actually had 1 or 2 branches in progress before stopping the project.. 

https://github.com/hoverinc/ray-tracing-renderer/tree/demodulate-albedo

[Here's LPL's API](https://lgltracer.com/docs/index.html#/api/LGLTracerRenderer?id=enabledenoise) using Temporal and Spatial options

**Comment by DennisSmolek** [comment]
Cuda implementation of SVGF with good notes

https://github.com/ZheyuanXie/CUDA-Path-Tracer-Denoising

**Comment by fraguada** [comment]
Hello! All I can add is my query in the open image denoiser repo back in January about a wasm compilation target: https://github.com/OpenImageDenoise/oidn/issues/129

While I work on a project where we compile a c code lib to wasm, I'm not sure I'd be able to handle the oidn source! 

Great that there are other alternatives to try!

**Comment by repalash** [comment]
How about starting with a simple bilateral shader.

**Comment by richardassar** [comment]
This is certainly doable and I do have deep learning experience having worked for a games industry middleware company, they applied deep learning to real-time speech animation synthesis from raw audio and there was often a lot of raw network surgery required there. I've also been actively researching and engineering in that space for many years, much more so in recent years, so I could definitely lend some help here.

The [model](https://github.com/OpenImageDenoise/oidn/blob/master/training/model.py#L48) is just a UNet, a fairly simple architecture that shouldn't present too many problems to port over to WebGL shaders. I think this issue should cover only that, leaving the support for generating our own training data as a separate issue.

An approach like [this](https://github.com/Erkaman/regl-cnn/blob/gh-pages/src/gpu.js) would work well, generating shader code parameterised by the layer configurations, possibly with some kernel fusion as things like ReLU can be fused into a previous convolution, for example.

OIDN does provide pretrained weights but these have likely been trained on a dataset that isn't representative of our output, also this misses the opportunity to improve the results by providing auxiliary input channels. 

Sampling a distribution of random scenes with various object placements, camera poses, material parameters etc. shouldn't be too tricky, and those can be POSTed off to a local endpoint which saves them out to disk for use in the training script. Alternatively the shaders in this project could be pulled into a native path tracer that invokes the main shader function from a compute shader, which is likely to fair better in terms of GPU occupancy and also isn't locked to VSync - this would allow us to scale to larger datasets and again is likely best split out into a separate issue.

If we want to provide extra input channels to the network we'll have to output them with multiple render targets, this could present an opportunity to experiment with the precise selection of what values to use in addition to things like first bounce, surface normals, etc. such as a per-pixel vector of material parameters and intermediate values from the whole path tracing computation that could be separately integrated/accumulated.

I propose that we bread it down as follows:

- Phase 1 (this issue): Load ODIN pretrained weights and perform inference on our "beauty" framebuffer, possibly providing some trivial auxiliary channels if they provide compatible weights
- Phase 2: Synthetic dataset generation within the browser
- Phase 3: Synthetic dataset generation with a native client sharing the shader code used by the web client

Those are my thoughts, I'm happy to start working on Phase 1.

**Comment by bhouston** [comment]
Interesting.  If we could start with pre-trained data, I think that is enough.  I would so much prefer to avoid having to train up our own data here, as we are not unique in terms of needs.  Given that Intel Open Source Denoiser is open source, couldn't we just use their weights?

**Comment by benzsuankularb** [comment]
RESTIR is another great option for this.
https://research.nvidia.com/publication/2021-06_restir-gi-path-resampling-real-time-path-tracing

**Comment by richardassar** [comment]
When we talk about input distribution this can also impact unexpected things like the camera projections, since ODIN is likely not trained on equirectangular projections this would be an example where the inputs lie on a manifold far from that which was sampled to produce training data. 

Since the network is learning a set of filter banks in the downsampling and upsampling convolution layers these end up to be specialized to the data, for example in raw audio networks you see gammatone filterbanks emerge for speech and very different filterbanks emerge when trained on non-speech. This is my concern, and things may still work fine, but as you push things you'll start to see unacceptable artifacts.

On the topic of denoising there are things like Diffusion Models and Inverse Autoregressive Flows (used in parallel Wavenet) that are also worth trying. One justification for putting infrastructure in place for importing models of various architectures and a synthetic data collection pipeline is that we can research and explore these, this will be important if we choose to send smaller models to mobile devices.

**Comment by richardassar** [comment]
[This trick](https://towardsdatascience.com/fourier-convolutions-in-pytorch-4cbd23c70005) might speed things up nicely, as it takes `O(n^2)` to `O(n*log(n))`.

**Comment by bhouston** [comment]
Very interesting.  If you want to add https://github.com/BrutPitt/glslSmartDeNoise @richardassar, which seems incredibly straight forward and with a high ROI, I can offer a bounty of $250 USD for that.  It seems relatively straightforward and a good first step. :)

**Comment by richardassar** [comment]
Investigating the feasibility of @benzsuankularb's proposal of implementing ReSTIR.

See https://github.com/lukedan/ReSTIR-Vulkan/blob/master/src/shaders/spatialReuse.comp

The lack of SSBOs (outside of WebGL2-compute) would mean replacing `reservoirs` with textures / a texture array and `resultReservoirs` with enough framebuffers to fit all of the `Reservoir` struct defined in https://github.com/lukedan/ReSTIR-Vulkan/blob/master/src/shaders/include/structs/restirStructs.glsl#L31 

The number of framebuffers required is `RESERVOIR_SIZE * 4 + 1`, which isn't terrible. `unbiasedReuse.glsl` also makes use of SSBOs but could be implemented with the same method.

This leaves the remaining challenge of figuring out how to integrate this with the existing three-gpu-pathtracer code.

**Comment by richardassar** [comment]
Submitted a PR for this, https://github.com/gkjohnson/three-gpu-pathtracer/pull/194.

The results seem reasonable but it requires a bit of parameter tuning to balance the tradeoff between denoising and excessive blur.

![Screenshot 2022-06-25 at 06-44-54 Material Orb Path Tracing](https://user-images.githubusercontent.com/868396/175760152-ff9f4c8f-1219-46de-9a08-ce8bd38e8e1f.png)

Give it a try.

**Comment by richardassar** [comment]
I saw improved results with the following settings, about double the `sigma` value and about one third of the `threshold`.

```
sigma = 5.32
threshold = 0.03
kSigma = 1.18
```

![Screenshot 2022-06-25 at 07-32-55 Material Orb Path Tracing](https://user-images.githubusercontent.com/868396/175761585-9c5f167d-fc1d-43e1-be7d-64793631552d.png)

Will push these. 

I also think that `radius` in the denoiser could be scaled by `resolutionScale`, either directly or by modifying the effective `sigma` and `kSigma` by `sqrt( resolutionScale )`, just to normalize things a bit.

**Comment by DennisSmolek** [comment]
> SSBOs but could be implemented with the same method.

I haven’t dived into your work (great job btw!!)

do you foresee a downside to using the “ reservoir” method over SSBO?

from my casual glance it looks like a dataTexture type setup, which even with the availability of attributes/buffers I’ve seen many modern/great libraries use,

Although it does feel like a hack..

**Comment by richardassar** [comment]
It's either a case of splitting things out over multiple render targets or one large one using a stride to split things into pages. I still need to take a look at what would be involved to integrate it into the rest of the path tracer. Next I'll try the ODIN (neural) denoiser and if people are interested then attempt ReSTIR at some point provided it looks feasible.

**Comment by gkjohnson** [comment]
Some more new denoiser updates and implementations:

- https://developer.nvidia.com/rtx/ray-tracing/rt-denoisers
- https://www.youtube.com/watch?v=W1UDzxtrhes

**Comment by DennisSmolek** [comment]
@gkjohnson What would be the best format to output from the denoiser?

I'm working on an OIDN to Tensorflow.js setup right now and trying to figure out the best way to take in/out the image data.

Tensorflow will auto convert all sorts of HTML image types but the data off the pathfinder is probably already a plain buffer.

I'm wondering too if I should return a NEW buffer or edit the buffer in place..

**Comment by thomasaull** [comment]
@DennisSmolek Just want provide the pointer to https://x.com/pissang1 who has implemented ODIN in (I think) threejs aswell and was mentioning somewhere he might open source it.

**Comment by DennisSmolek** [comment]
> @DennisSmolek Just want provide the pointer to https://x.com/pissang1 who has implemented ODIN in (I think) threejs aswell and was mentioning somewhere he might open source it.

Thanks!
[We actually talked about it.](https://x.com/DennisSmolek/status/1805163520155414999)

He confirmed some of my thoughts on execution but he doesn't seem up to release just yet.

[I also discussed it](https://github.com/Twinklebear/oidn-rs/issues/30#issuecomment-2186749015) with the OIDN Rust wrapper dev to use Tensorflow.js vs the CPU wrapper.

With the WebGPU backend being so fast and the ability to keep the buffer on the GPU I'm inclined to agree.

**Comment by thomasaull** [comment]
@DennisSmolek I see, way ahead of me :) Would definitely be a game changing feature for three-gpu-pathtracer :)

**Comment by gkjohnson** [comment]
@DennisSmolek Really excited to hear you're looking into this. I was hoping someone might be interested in it since I saw Max Liani's work.

I'm less familiar with Tensorflow and how it works so feel free to correct any of my assumptions.

> I'm working on an OIDN to Tensorflow.js setup right now and trying to figure out the best way to take in/out the image data.

I'm not sure what other restrictions there might be but I see that Tensorflow supports both a WebGL and WebGPU backend. Is the TF interface backend agnostic? Ie can you easily run the same model with either backend without too many code changes? Without knowing too much in depth I'm imagining passing the pathtraced WebGL frame buffer to a function and model outputs the denoised result into a different WebGL framebuffer.

If it requires the WebGPU backend, though, then we'll have to read the WebGL framebuffer to the CPU and upload it to the WebGPU context and display it afterward. Some day this project will move to WebGPU, too.

> I'm wondering too if I should return a NEW buffer or edit the buffer in place.

Lets allow for writing into a separate buffer since I think a common case would be progressively updating the "noisy" path traced buffer every frame and then denoising into a separate buffer for display. But I think whatever works for getting a first implementation going is best and then we can adjust afterward.

Let me know if there's anything else I can do to help with this! Excited to see any updates, as well

**Comment by DennisSmolek** [comment]
> I'm not sure what other restrictions there might be but I see that Tensorflow supports both a WebGL and WebGPU backend. Is the TF interface backend agnostic? Ie can you easily run the same model with either backend without too many code changes? Without knowing too much in depth I'm imagining passing the pathtraced WebGL frame buffer to a function and model outputs the denoised result into a different WebGL framebuffer.
> 
> If it requires the WebGPU backend, though, then we'll have to read the WebGL framebuffer to the CPU and upload it to the WebGPU context and display it afterward. Some day this project will move to WebGPU, too.

In general, tensorflow is very backend agnostic allowing you to fallback based on system support or usage. For the javascript side of things the code for all backends is identical, so no issues there switching around. Handling the output is where it matters.

[In this section of the docs](https://github.com/tensorflow/tfjs/blob/master/docs/OPTIMIZATION_PURE_GPU_PIPELINE.md) they actually show how to use both the WebGL and WebGPU backends and pass the buffers to however you want to process it.

[There are examples as well](https://github.com/tensorflow/tfjs-examples/tree/master/gpu-pipeline)

The webGPU Backend is supposed to be even faster than the WebGL one, however as the pathtracer is using the WebGL backend it may be worth it to lock into the WebGL backend if it can keep the resulting buffer on the GPU.
( I'm not sure thats possible, I'm a little weak on compute with WebGL2.)

The other thing is it shows how to keep the data OUTPUT on the GPU but not the data INPUT. So I think regardless of backends we will have to take the framebuffer and input it into the denoiser on the CPU side.

Then the other optimization functions like upscaling can run (I noticed the "render size" option so I'm assuming thats what is happening)

I think with the current WebGL backend the ideal flow would be:

1. Render the noisy framebuffer (smallest resolution setting) return to the CPU
2. Send the framebuffer data to the denoiser using the WebGL backend
3. The denoiser runs and passes back the `GPUData` reference
4. The next steps in the render chain take place (upscale, tonemap, whatever)


I'm not sure what the limits are exactly with WebGL, however I know with WebGPU the buffer can persist on the GPU between draw calls and compute executions, so in the future there will be even less CPU round trips. [I've asked](https://github.com/tensorflow/tfjs/issues/8314) already about the possibility of executing the tensorflow initialization on the compute side too but its just a feature request at the moment.

---

For now, I have a basic Autoencoder denoiser running to help me setup the tensorflow core. I'm working on loading/converting the OIDN weights and setting up their much more complex UNet model now. I hope to be done by the end of the week with a basic version. I'll probably use the pathtracer as my real-world test! I'll be sure to update with my results.

**Comment by gkjohnson** [comment]
> - Render the noisy framebuffer (smallest resolution setting) return to the CPU
> - Send the framebuffer data to the denoiser using the WebGL backend
> - The denoiser runs and passes back the GPUData reference
> - The next steps in the render chain take place (upscale, tonemap, whatever)

Thanks for the detailed response! I can't dig in too deeply right now so I'll have to rely on your expertise on this but this seems good to me. Is there a specific "upscale" step you're referring to from OIDN, though? Or is this just an example?

> WebGL backend it may be worth it to lock into the WebGL backend if it can keep the resulting buffer on the GPU.

If it's not significantly different then this seems best to me - then we can just draw the resulting framebuffer to the canvas.

**Comment by gkjohnson** [comment]
> I saw some of the pathtracer demos have a "render" parameter that is usually ~0.5. It doesn't change the scale of the canvas but ups the load if you set it to 1.

Oh right - yeah this is just a basic linear interpolation upscale when drawing to the canvas. There's nothing fancy happening there.

**Comment by DennisSmolek** [comment]
So something small to add, Reading [the TensorFlow Docs](https://js.tensorflow.org/api/latest/#tensor) I found the section on creating the tensor from GPU data. according to this, we could indeed maintain the data completely on the GPU as inputs.

We'd still need to return to the CPU to initialize the call, but the data itself can stay on the GPU as a gl.texture. In WebGPU it's much nicer and cleaner as it would be a plain buffer. 

The only problem with this method at the moment is to get the best results for OIDN you would pass the standard color input (the noisy image) and then also pass a Albedo and normals image. 

Right now I handle this using tensorflow to concat the data as a part of the model execution. I'm not sure if this operation is CPU bound or not... 

If it were a single input buffer (just the color pass) I don't have to concat it, but it's said the extra data greatly improves the quality. Something I'll have to test.


Status wise I'm going slow as its a huge pain to convert both the python and C++ logic to Typescript.
I have the models loading, and have the weights, but I havent gotten a running test yet. 

Max Liani's work is actually using a 1.X branch and is out of date, plus he never actually says how he got the model working (if he even did?)

I'm sticking with plain image data for the moment, once I have that working I'll cleanup the library and then start messing with the GPU buffers.

**Comment by gkjohnson** [comment]
> If it were a single input buffer (just the color pass) I don't have to concat it, but it's said the extra data greatly improves the quality. Something I'll have to test.

I'd be curious as to what Max Liani and Yi Shen used for their output. It's a very different approach but the SVGF denoiser required this but seemed like it wouldn't behave well with transparent objects. OIDN is probably better in this case. If needed we can look into writing these buffers out, as well, using MRT. But as you mention I say lets start with just the noisy path traced buffer to start.

**Comment by gkjohnson** [comment]
Wanted to add this here so I don't forget it - but it sounds like using white noise during path tracing will provide a nicer result with OIDN than the current blue-noise stratified samples used currently:

https://x.com/pissang1/status/1803019554177843354

<img width="587" alt="image" src="https://github.com/gkjohnson/three-gpu-pathtracer/assets/734200/35732e69-bb65-4fb1-a0c8-c202f9a14a6b">

**Comment by DennisSmolek** [comment]
> Wanted to add this here so I don't forget it - but it sounds like using white noise during path tracing will provide a nicer result with OIDN than the current blue-noise stratified samples used currently:
> 
> https://x.com/pissang1/status/1803019554177843354
> 


Is that something easy to change/adjust?

Those results look great! We need to make sure people understand that for that level of clarity you likely need the Albedo and Normals. The docs push this as the best results, and Yi Shen confirmed that all his use those layers as well. 
Slight MRT work shouldn't be too bad.

I haven't tested it yet but my setup is already designed to take those layers automatically.

Oh and in further news, I got OIDN working!
https://twitter.com/DennisSmolek/status/1807814362851401828

![Screenshot 2024-07-02 020325](https://github.com/gkjohnson/three-gpu-pathtracer/assets/1397052/6c28c153-8fc6-4cef-88c6-52a9ab670e1c)

I haven't shared this on twitter yet but talking with the tensorflow folks I learned that the first pass through the WebGL backend should be discarded as it takes that pass to setup the shaders/backend.

For a 720 X 1280 image I went from 2,000ms - 15,000ms to predict and display to 4ms-7ms with WebGL.

The reason WebGPU was initially faster was it doesn't have as long of a shader/setup phase.

They actually said WebGPU so far is usually the same speed or slower in many cases and only a few cases faster. 
Definitely something to test.

Right now the library is taking in any ImageData, Array, or HTMLImageObject as input and can output in the same ways. 
Today I'm working on the buffer stuff which I don't anticipate a lot of problems with. 

I'm going to use the lego example as I actually have a few of the sets so it'll be interesting.

As soon as I have the buffer working I'll publish the repo for feedback.

**Comment by DennisSmolek** [comment]
So something to note...

I have my denoiser library able to take in an external webGL context to use for the tensorflow.js webgl backend,  so the texture output can be shared with other applications (like threeJS)

Problem is, Even connecting to a running threeJS context seems to have conflicts between the two applications.. I think because of location binding.
This part is getting into the GPU weeds a bit for me,  but I think for at least 1.0 of the denoiser texture syncing isn't going to work until the conflict can be resolved.

![image](https://github.com/gkjohnson/three-gpu-pathtracer/assets/1397052/760c8ec7-1e79-4750-8877-e372fcea3804)

I spent a day setting up gl in/out and I'm not sure how to fix this. 

Also, ThreeJS REALLY doesn't want you to pass in your own `WebGLTexture` to texture objects, but I wasn't able to finalize testing on this as the two systems dont like sharing the same context at all. 

This also means direct GPU Input isn't going to work either, as that also requires GPU context sharing.


I'm more hopeful for WebGPU in this regard, but as the pathtracer is webGL, it doesn't help much. 

For now, I'm going to back burner GPU Input/Output and just do a CPU Sync.

Pathtracer renders to a framebuffer, pass to the denoiser, denoiser outputs an imageData buffer that I push on a texture and back onto the threeJS backend.

some reading says I might be able to speed this up with an Offscreen Canvas...

**Comment by repalash** [comment]
Sorry to jump in the middle here, but isn't it possible to render to a canvas and use in threejs as a CanvasTexture

**Comment by DennisSmolek** [comment]
> Sorry to jump in the middle here, but isn't it possible to render to a canvas and use in threejs as a CanvasTexture

Yeah that's totally possible, and might be what I do if I use the offlineCanvas

The thing is to render onto the Canvas it has to take the data off the GPU and convert it on the CPU to a Buffer. Then we draw it onto the canvas.
Then threeJS Takes that canvas, makes it into ANOTHER buffer, and sends it back to the GPU in a material. 
Which then of course comes back off the GPU to the final canvas render.

Tensorflow outputs as a texture, meaning, (if there were no conflicts) the data never leaves the GPU and skipping all the CPU buffers and straight to the last step of rendering.

It's the same with the input. Tensorflow accepts a framebuffer/texture as Input, so you could draw to the framebuffer and that data would stay on the GPU, get processed, and never leave. This saves a lot of speed/bandwidth.

It's not a dealbreaker or the end of the world, but it sucks knowing that capability is there just not usable.

**Comment by gkjohnson** [comment]
cc @mrdoob

tldr; Dennis is working on integrating an ML denoiser using [tensor flow](https://www.tensorflow.org/) and we'd like to use the same WebGL context from three.js' WebGLRenderer to avoid copying data between contexts. However three.js seems to break when sharing the context due to state inconsistency. It also seems not possible to get or manually provide frame buffer handles so they can be shared with tensor flow. Are there any recommendations here? (see https://github.com/gkjohnson/three-gpu-pathtracer/issues/85#issuecomment-2205467554)

@DennisSmolek correct me if I've misunderstood anything. Calling [WebGLRenderer.resetState](https://threejs.org/docs/?q=webglrendere#api/en/renderers/WebGLRenderer.resetState) before and after doing anything with TensorFlow may help but if TF requires that no state changes happen between multiple frames (ie running the model and reading) then it may still be an issue.

> For now, I'm going to back burner GPU Input/Output and just do a CPU Sync.

Seems like a good plan for now. I'm not as familiar with WebGPURenderer, yet, so it's not clear to me whether these things will still be an issue with WebGPU, unfortunately.

**Comment by mrdoob** [comment]
> @DennisSmolek correct me if I've misunderstood anything. Calling [WebGLRenderer.resetState](https://threejs.org/docs/?q=webglrendere#api/en/renderers/WebGLRenderer.resetState) before and after doing anything with TensorFlow may help but if TF requires that no state changes happen between multiple frames (ie running the model and reading) then it may still be an issue.

I was going to say exactly this 👀

**Comment by Twinklebear** [comment]
I saw that Yi Shen just open sourced their web port of OIDN which might provide some inspiration (or maybe can just integrate that) @DennisSmolek : https://x.com/pissang1/status/1809119875258011671

**Comment by DennisSmolek** [comment]
> I saw that Yi Shen just open sourced their web port of OIDN which might provide some inspiration (or maybe can just integrate that) @DennisSmolek : https://x.com/pissang1/status/1809119875258011671

Yeah... A bit confused by that as I asked about it and didn't think he was going to open source it.
Glad it's out there though! Just wish I knew that was his plan a few weeks ago when we first chatted.

Unfortunately his library doesn't support webGL GPU handling which is the current blocker.

Because we both pull from OIDN our UNets are the same but much of the rest of our code is different.

I wanted to wait until I had webgl working to release but I don't want people to loose interest so I'll be releasing my current state with my webgl testing app so hopefully other people can help.

I'll work on docs and an example and get it up tomorrow or Sunday

**Comment by DennisSmolek** [comment]
Just to clarify...

I have GPU side input and output working for webGL and for WebGPU.

However, (in WebGL) threeJS specifically is having context conflicts with tensorflow.

This isn't an issue with WebGPU as the buffers aren't limited to the same canvas/state, just the same device.

My ultimate goal with denoising has always been this pathtracer, and with it being WebGL bound my #1 goal is to get ThreeJS and the denoiser to work together to pass data/textures on the GPU using webGL.

**Comment by gkjohnson** [comment]
> However, (in WebGL) threeJS specifically is having context conflicts with tensorflow.

To be clear calling `renderer.resetState` isn't helping here? Is there state that _isn't_ being reset by three.js correctly that could be fixed? Or is tensor flow expecting state to be remain unchanged asynchronously for a longer duration?

> This isn't an issue with WebGPU as the buffers aren't limited to the same canvas/state, just the same device.

Interesting - this is good to know.

If it looks like it won't be possible to make any changes to three.js to address the WebGL issue then a WebGPU implementation seems like a fine substitute even if it's not ideal.

**Comment by DennisSmolek** [comment]
The code isn't perfect and I'm not ready to blast it out yet, but I have running demos of my denoiser and the rough first draft of the docs live.

[Live WebGPU Demo Here](https://webgpu-basic.vercel.app/)
Code: https://github.com/DennisSmolek/Denoiser

having vercel issues so I will have to push live demos manually once things get cleared up.

I hope to resolve my webGL issues soon then I'll focus on three/webGL and the pathtracer.

Please take a look at the demo/docs and let me know what you think.

I'll push out on twitter after a few cleanup passes, hopefully have things resolved.

**Comment by DennisSmolek** [comment]
> > However, (in WebGL) threeJS specifically is having context conflicts with tensorflow.
> 
> To be clear calling `renderer.resetState` isn't helping here? Is there state that _isn't_ being reset by three.js correctly that could be fixed? Or is tensor flow expecting state to be remain unchanged asynchronously for a longer duration?

Honestly, I'm not sure. I put things on pause to get the repo up so I'll be readdressing. Essentially, when running in a threejs shared environment the model returns all 0's and no errors were thrown. 

I tried multiple versions/setups of `resetState` as well as saving/restoring the state and program for tensorflow with Cody writing the save/restore state snippets.
I'm working on a vanilla WebGL version and once I have that working I'll try again on the threeJS WebGL version.
Now that it's public too when I run into errors it will be easier to share.

I honestly wonder if running OpenGL backend in chrome to speed up the pathtracer compile time (im on NVIDIA GPU) could also be playing a part.


> If it looks like it won't be possible to make any changes to three.js to address the WebGL issue then a WebGPU implementation seems like a fine substitute even if it's not ideal.
If you mean copy the output from three/webGL and move it to WebGPU this will totally work, however the CPU/GPU sync to get the data sucks. Like on average 1600ms.

It would be the same to go from three/webGL to the denoiser/gl and probably work (meaning devices that dont support webGPU could use it) but it would be slow...

The pure webGPU version is significantly more stable and easier to share buffers around. So once the pathtracer is on WebGPU I think it will be even better.

**Comment by DennisSmolek** [comment]
I wanted to tell everyone following in this thread first,
[ThreeJS/WebGL Version up and running!](https://denoiser-threejs-webgl.vercel.app/)

![Screenshot 2024-07-13 010811](https://github.com/user-attachments/assets/6b1ab9f3-f355-488e-8d73-c948658b4d5a)

I had to write an entire `WebGLStateManager` inside the denoiser on top of using `renderer.resetState()`

Plus getting raw `WebGLTextures` in/out of three is undocumented and has changed over the decade+ of threejs.

Checkout [The Demo](https://denoiser-threejs-webgl.vercel.app/) and lmk if it doesn't work or acts weird.

I'll be spending Saturday afternoon finishing up the docs and maybe trying to add more examples. Then I'll blast in out on SNS.


We've lost some speed to the tiling but as first drafts go I'm happy with it.

**Comment by thomasaull** [comment]
It's a small detail, but the demo image shifts a bit between the original and denoised version

**Comment by DennisSmolek** [comment]
> It's a small detail, but the demo image shifts a bit between the original and denoised version

Yeah I’ll fix it. The data doesn’t really shift, it’s just the canvas flex/position is absolute while the preview is three images in a container.

**Comment by thomasaull** [comment]
@DennisSmolek A quick fix would be to add the following changes to `.imgWrap`:

- Remove `width`
- Add `flex-basis: 100%`

**Comment by gkjohnson** [comment]
Very nice! 

> I had to write an entire WebGLStateManager inside the denoiser on top of using renderer.resetState()

I'm wondering if this is something that could be adjusted in three.js? Sounds like `resetState` may be missing some things? Unless the tensor flow package is expecting the state remain to unchanged, as well.

> this isn't the best example on speed as the webGL version has to compile the first time it runs

I'm not sure if this is something you'd have control over in TF but the [KHR_parallel_shader_compile](https://developer.mozilla.org/en-US/docs/Web/API/KHR_parallel_shader_compile) could be able to help here. The page won't be blocked so you can compile on page load and be ready for denoising asap. Though I didn't see any compile time the performance trace so maybe this is already happening (granted I'm not on Windows).

> The "Denoise" time is also misleading as with GPU being asnyc we cant be 100%.

With the [EXT_disjoint_timer_query](https://developer.mozilla.org/en-US/docs/Web/API/EXT_disjoint_timer_query) you should be able to get more precise GPU timing.

You may already know about these extensions but I thought I'd mention them.

**Comment by DennisSmolek** [comment]
So I'm about to post my version with the pathtracer.

Something is weird with the way it's handling output from the pathtracer.

First I think the pathtracer renders to it's renderTarget in HDR. This isn't a problem but I'd appreciate if someone could confirm it.

Second, either the blue noise/white noise thing[ mentioned here ](https://github.com/gkjohnson/three-gpu-pathtracer/issues/85#issuecomment-2201832953) or something else in the way the pathtracer coalesces is causing issues.
@gkjohnson You mention stratification, is that something that could be causing the issue?

Here is a 6spp image that fails the denoiser:
 
![Screenshot 2024-07-17 215907](https://github.com/user-attachments/assets/08892a8a-0d40-440a-98b1-386cfd51e225)

Here is a 4spp one that passes:
![K68HcTb](https://github.com/user-attachments/assets/188373c4-fa8c-404a-82bc-809459dd8c16)

[The OIDN Docs]() mention:
>The RT filter has certain limitations regarding the supported input images. Most notably, it cannot denoise images that were not rendered with ray tracing. Another important limitation is related to anti-aliasing filters. Most renderers use a high-quality pixel reconstruction filter instead of a trivial box filter to minimize aliasing artifacts (e.g. Gaussian, Blackman-Harris). The RT filter does support such pixel filters but only if implemented with importance sampling. Weighted pixel sampling (sometimes called splatting) introduces correlation between neighboring pixels, which causes the denoising to fail (the noise will not be filtered), thus it is not supported.

I've seen in other places people mention reflections/speculars that the denoiser rejects. So what I think is happening is the denoiser doesn't want to display black for missing spaces, so is either blending with the standard forward pass or is bluring in nearby data for the missing spots. 
The denoiser reads that as having image data and ignores it.

Here is a single image with my test output from the pathtracer, and the OIDN test image. 
When ran, it denoises the bottom and ignores the top:
![merged](https://github.com/user-attachments/assets/2ff21ed8-422e-411e-8780-dfb6903c3758)


I thought my denoiser was just broken, but I've since tested 20 different pre-OIDN images of many quality sizes and renderings and all of them work.

[Max on twitter](https://x.com/MaxGaniyev/status/1812623379121287239) is using his own pathtracer with my denoiser and getting great results. (SUPER fast too, and he's using the base WebGL to canvas version)

Yi Shen's test images also run through mine just fine.. 

Is there some hidden setting/option I can adjust to change the pathtracers behavior?

**Comment by gkjohnson** [comment]
> First I think the pathtracer renders to it's renderTarget in HDR. This isn't a problem but I'd appreciate if someone could confirm it.

The path tracer renders to a float 32 buffer and basically stores light intensity in linear color space which will often be larger than 1. Then when rendering to the canvas the linear colors are tone mapped to sRGB for display.

> Second, either the blue noise/white noise thing mentioned here or something else in the way the pathtracer coalesces is causing issues.
> @gkjohnson You mention stratification, is that something that could be causing the issue?

It might be easier to help once I can see what other changes you've made to the pathtracer settings already or what the set up is - but if you're just using the default settings then yes it will be using stratified sampling with a blue noise offset. It's harder to tell from the tunnel image but the lego images are definitely using the stratified sampling.

You can swap to use white noise by changing the [RANDOM_TYPE define value to 0](https://github.com/gkjohnson/three-gpu-pathtracer/blob/7e7be065187272486cd9aa12abc66e213c367e8f/src/materials/pathtracing/PhysicalPathTracingMaterial.js#L53-L56) (PCG randomness).

It also looks like the images you're trying to denoise are being rendered at half-resolution (or some resolution less than 1) which may also be causing issues.

Regarding anti aliasing - I'll have to check on the filter again when I time but if the above issues don't help you can disable AA by changing [this line](https://github.com/gkjohnson/three-gpu-pathtracer/blob/7e7be065187272486cd9aa12abc66e213c367e8f/src/materials/pathtracing/glsl/camera_util_functions.glsl.js#L16) to `vec2 jitteredUv = vUv;`.

**Comment by DennisSmolek** [comment]
> The path tracer renders to a float 32 buffer and basically stores light intensity in linear color space which will often be larger than 1. Then when rendering to the canvas the linear colors are tone mapped to sRGB for display.

Ah ok. I'm already setup to convert from linear if needed. Will that bring it down to a normalized range or could it still be above 1? If using real light intensity I can adjust with a scalefactor to get the HDR in expected range but I haven't had to do that yet.

My pathtracer setup is very simple:
```ts
this.pathtracer = new WebGLPathTracer(this.renderer);
this.pathtracer.setScene(this.scene, this.camera);

// disable default fade and protections
this.pathtracer.renderToCanvas = false;
this.pathtracer.renderDelay = 100;
this.pathtracer.fadeDuration = 0;
this.pathtracer.minSamples = 0;
this.pathtracer.renderScale = 1;
// recent adjustment
this.pathtracer.multipleImportanceSampling = true;
```

Is there a way to adjust that material without using my own build of the pathtracer? I can pull and build my own if I need to but I was trying to avoid it in my demo.

One of those samples I was indeed using a lower render scale as a test, but this happens regardless.

I've looked at Y- Shen's version. He outputs the pathtracer to a canvas then copies that as image data into his denoiser. I could try this too but honestly I've not been able to replicate the results from his video when using the demo. Does it work better on your mac?

**Comment by DennisSmolek** [comment]
Ok tested on my mac both Yi-Shen's and my denoisers are significantly faster there and work.
On windows, his pathtracer demo does not work, it just outputs black tiles. On mac it works very quickly which is nice.

I uploaded my (currently broken) pathtracer demo to Vercel. This seems to have fixed some things but let others still be broken.

you can see it here: https://denoiser-three-pathtracer.vercel.app/

The blown out areas make me think my handling of the input tensors is wrong and I need to adjust them. Either SRGB or some other setup.

But the noise is mostly gone  .. 🤷‍♂️

This demo isn't totally stable or setup right so I wouldn't publish it.

**Comment by gkjohnson** [comment]
> you can see it here: https://denoiser-three-pathtracer.vercel.app/

Is there anywhere to see the code?

> The blown out areas make me think my handling of the input tensors is wrong and I need to adjust them. Either SRGB or some other setup.
>
> But the noise is mostly gone .. 🤷‍♂️

Yeah the noise looks improved but it looks like inputs or something look like they may be off?

_From before the edit since I already had it typed_

> Ah ok. I'm already setup to convert from linear if needed. Will that bring it down to a normalized range or could it still be above 1?

It depends on how you're converting from linear and what to. If you mean you're converting to sRGB color space then yeah that would typically mean your values are in the range [0, 1]. I'd typically do this by using a full screen quad with a MeshBasicMaterial to handle the tonemapping to a new render target.

> Is there a way to adjust that material without using my own build of the pathtracer? I can pull and build my own if I need to but I was trying to avoid it in my demo.

Not officially supported (we can adjust support later) but you can set the noise style with the following flag:

```js
const pathTracer = new WebGLPathTracer( renderer );
pathTracer._pathTracer.material.defines.RANDOM_TYPE = 0;
```

**Comment by DennisSmolek** [comment]
Latest update: https://denoiser-three-pathtracer.vercel.app/

This version is 100% GPU based from three, to the pathtracer, to the denoiser, back to three for final rendering.
SIGNIFICANTLY faster on my Mac which was already fast, 1000-2000ms on my PC but it's hard to tell because my PC averages 5-8,000ms in general.

Colors look weird IMO. And be sure to use "balanced" quality for the best quality check. Fast can be wonky.

The denoiser is running as it should but my demo is still struggling to figure out the right colorspace/tone settings of things. The colors are coming in brighter/whiter than they should and I think it's messing things up.

I'm also not 100% confident on my albedo & Normals generation

The code for this example is [Here](https://github.com/DennisSmolek/Denoiser/blob/main/examples/three-pathtracer-texture-in/)
It's super messy and convoluted. [Here is the quad/mixing shader](https://github.com/DennisSmolek/Denoiser/blob/62297efd6d278051c6f6041bba253379fab1d5ac/examples/three-pathtracer-texture-in/src/threePathtracerDemo.ts#L217) in the rest of it.

How the renderer works.

-  Load Model and setup basics
- Setup main scene and fullScreenQuad Scene
- Setup pathtracer and denoiser
- Setup Control panel

Then the render loop:

1. Warm up special render target for denoiser (initializes texture we later inject into) only runs first loop
2. Run special Albedo/Normals MRT pass to generate Aux's for later
3. Render Pathtracer (stop at maxSample)
4. Render special blend fullscreen quad which has all textures and blends accordingly.
5. If pathtracer samples is the set denoise target (default 6), run Denoiser
6. Denoiser takes pathtraced texture, applies it to our standard fullscreenQuad and renders it out to another texture. (should be SRGB, Probably regular linear)
7. Extract Raw WebGLTexture from textures, set color, albedo, and normals on denoiser
8. Execute Denoiser
9. Get denoiser output, Inject raw WebGL into Texture object (probably should be SRGB output, probably is Linear)
10. Now that denoiser texture is set, next frame the blender (step 4) will correctly blend the denoised image over everything.

How the blender works:
1. Start with base three.js render of the scene.
2.  Render pathtracer
3. When pathtracer reaches minimum Samples use timer to start a blend between base render and pathtraced
4. When denoise trigger number of samples reached start denoiser
5. When denoiser results come back, blend between denoised and pathtraced
6. Any user input or certain options changes, immediately reset the pathtracer, reset all blends and show base threejs render
7. Repeat


All of the code is live, I'm 100% open to changes and criticisms. I know things need to be cleaned up and I'm totally open to suggestions for the main package or any example.

One thing to note is I will likely be moving the repo to Poimandres. They've agreed to take the project into the collective which will hopefully get more eyes on it to help it improve.

The NPM `denoiser` will probably stay the same because its so short.

**Comment by bhouston** [comment]
This is amazing!  We are close!  My bounty is still open BTW!

I think that there is a brightness issue?  Compare the pathtraced with the denoised and the denoised is definitely all around brighter.  I think that it should somehow preserve the overall intensities (e.g. the histogram should stay the same between the path traced and denoised) rather than brightening them.

<img width="1305" alt="Screenshot 2024-07-21 at 9 09 37 PM" src="https://github.com/user-attachments/assets/38ac4bea-1fed-44ca-9696-4b4a6d5e25e9">


<img width="1305" alt="Screenshot 2024-07-21 at 9 09 42 PM" src="https://github.com/user-attachments/assets/0ee144ac-6b4b-46ad-b6d0-34315d8d0286">

**Comment by DennisSmolek** [comment]
> I think that there is a brightness issue? Compare the pathtraced with the denoised and the denoised is definitely all around brighter. I think that it should somehow preserve the overall intensities (e.g. the histogram should stay the same between the path traced and denoised) rather than brightening them.

Totally agree.

Part of me thinks it's something to do with colorSpace another thinks it's from my albedo calculations. If you look at the original "color" compared to the "albedo" pass (which is essentially just a `basicMaterial` you can see the over bright color.

[The OIDN Docs](https://github.com/RenderKit/oidn?tab=readme-ov-file#albedos) say:
> The albedo image is the feature image that usually provides the biggest quality improvement. It should contain the approximate color of the surfaces independent of illumination and viewing angle.
> For simple matte surfaces this means using the diffuse color/texture as the albedo. For other, more complex surfaces it is not always obvious what is the best way to compute the albedo, but the denoising filter is flexible to a certain extent and works well with differently computed albedos. Thus it is not necessary to compute the strict, exact albedo values but must be always between 0 and 1.

And goes on to explain more in detail.

If I set the debugging flag `usePassthrough` it creates a bypass version of the model (to test input/output handling) and it returns almost the exact pathtraced input, so it's for sure something happening withing the denoising flow.

But if I use other OIDN test images I get the expected results. So I'm fairly confident it's just something in how I'm putting the data in/expect the results

**Comment by DennisSmolek** [comment]
So I've improved my Albedo and normals pass including the ability to use worldNormals:
![image](https://github.com/user-attachments/assets/bfdf9cef-cfaf-4a88-86fd-9076b0a8ae7c)

I've also since learned that OIDN CAN accept SRGB as input with a flag but expects linear by default. Meaning the issue is likely not one about linear in/out though I'll keep testing.

Reading about brightness/colors makes me think the issue is something with tonemapping. Either the pathtracer output into the denoiser needs tonemapping applied or the same with the output of the denoiser.

I've also made some ground on speed improvements and adding better logging/debugging.
There is a balance between speed/memory. The less batches the faster it runs but the bigger the memory draw.

There is a calculateBatchSize function that I haven't resolved the logic yet besides `if (isMobile) return 1` which keeps the memory down the most but makes it the slowest.

I have also found we can force tf to use float16 textures for webGL which (may) make it faster. But will def help with mobile issues and may be why older models were crashing despite the lower memory reqs.

There are 4-5 other speed improvement ideas I have ranging from super experimental to absolutely will work but is super hard to execute.

Right now goals are:
1. the Tonemapping/brightness
2. Transparency 
3. Speed on Desktop
4. Mobile stability

**Comment by gkjohnson** [comment]
Thanks for putting all this together @DennisSmolek!

I've taken a brief look at the code and these are some things that stick out to me:

- Albedo texture is perform [4x4 samples](https://github.com/DennisSmolek/Denoiser/blob/62297efd6d278051c6f6041bba253379fab1d5ac/examples/three-pathtracer-texture-in/src/albedoAndNormalsRenderPass.ts#L66-L80) of the texture map which will cause a blurry texture surface.
- Looking at the [Linear sRGB path traced texture conversion to sRGB](https://github.com/DennisSmolek/Denoiser/blob/main/examples/three-pathtracer-texture-in/src/threePathtracerDemo.ts#L433-L443) - it looks as though [no conversion to sRGB is performed](https://github.com/DennisSmolek/Denoiser/blob/62297efd6d278051c6f6041bba253379fab1d5ac/examples/three-pathtracer-texture-in/src/threePathtracerDemo.ts#L283-L288). The Albedo texture, however, is rendered to sRGB so there's a mismatch between the color spaces and value ranges of the albedo and path tracer input.
- From other discussions on twitter it seems like we should be using white noise for this denoising rather than the current stratified noise.
- The rendered normal buffer doesn't use any normal maps.
- Small nit but `WebGLRenderer.setPixelRatio` should be called so the render is using the correct device pixel resolution.

The brightness differences may come from color space inconsistencies as you mention.

And lastly it should be figured out what the right way is to render the albedo and normal textures - specifically regarding transparent surfaces and depth of field. It's possible these textures are supposed to be averaged results in the same way the path traced input texture is. This would require changes to the path tracer, of course. From other discussions it sounds like switching to use the white noise setting in the path tracer may help the results, as well. 

---

Whenever someone wants to make a PR into this project to add support for one of the OIDN pathtracers to `WebGLPathTracer` that would be great. We can start with the three.js rasterized albedo and normal textures and improve over time.

**Comment by DennisSmolek** [comment]
Made some changes and have some results: https://denoiser-three-pathtracer.vercel.app/
![Updated](https://github.com/user-attachments/assets/9a0f34da-c551-4614-9e13-847ac8b1f1ab)
On the left is what was up until a few minutes ago, on the right are with my changes. 

From what I can tell, in general, the more samples the closer to ground truth.

Minor color space changes noticeable even in the pathtraced results.
With the older version the bad results get magnified and blow things out.
Newer version blends much closer to the original PT results.

At 100 samples, the blown out results lessen and it gets much closer to the PT source. However the newer version looks pretty consistent (although naturally better) to the 6 sample and PT version. 

These are both with "fast" quality on purpose which gets the worse results but helped show the changes. 

Play with it and lmk. be sure to up the samples and "denoise at"

Negatives: I did notice some roughness in dark edges. I'm not sure of the source. I think it may be the new normals. Will add a test to go between the worldspace and viewspace.

**Comment by DennisSmolek** [comment]
> * Albedo texture is perform [4x4 samples](https://github.com/DennisSmolek/Denoiser/blob/62297efd6d278051c6f6041bba253379fab1d5ac/examples/three-pathtracer-texture-in/src/albedoAndNormalsRenderPass.ts#L66-L80) of the texture map which will cause a blurry texture surface.

I did this because docs said inputs should be anti-aliased the same. I didn't notice any bluryness in the albedo output (you can change the output and see) If you think it best raw we can totally remove it.

> * Looking at the [Linear sRGB path traced texture conversion to sRGB](https://github.com/DennisSmolek/Denoiser/blob/main/examples/three-pathtracer-texture-in/src/threePathtracerDemo.ts#L433-L443) - it looks as though [no conversion to sRGB is performed](https://github.com/DennisSmolek/Denoiser/blob/62297efd6d278051c6f6041bba253379fab1d5ac/examples/three-pathtracer-texture-in/src/threePathtracerDemo.ts#L283-L288). ~The Albedo texture, however, is rendered to sRGB~ (I'm actually not exactly sure what color space the Albedo textures are in since a RawShaderMaterial is being used) so there's a mismatch between the value ranges and (possibly) color spaces of the albedo and path tracer input.
I'm not sure right this second TBH. It's the default so I thought it was coming out as Linear. I will check though.

I have made adjustments to the Albedo and Normals pass to add the worldspace normals option (to match the OIDN examples) although OIDN says either world or viewspace is acceptable. I noticed some darkness/oddness in the edges though so I think this a big area of attention.

> * The rendered normal buffer doesn't use any normal maps.

Like sampling from a provided map? I generated the normals at RT but sampling a map if it has one makes sense. 

> And lastly it should be figured out what the right way is to render the albedo and normal textures 

Yeah, I just tried my best here and tried to keep it simple. It's straightforward to get them with threejs forward passes but (especially with transparency) may need more advanced versions. Totally open to whatever yall think best.

Sidenote, If we use the pathtracer for the Aux Inputs its highly recommended to pre-pass denoise them before sending to the final denoiser.

> Whenever someone wants to make a PR into this project to add support for one of the OIDN pathtracers to `WebGLPathTracer` that would be great. We can start with the three.js rasterized albedo and normal textures and improve over time.

Do you have any opinions on API's/presets?
Do you want to assume the pathtracer will be generating the albedos and normals by default or leave that as an external object the user has to setup/use? 

Do you want the main render call to use the denoiser automatically?

I was thinking something like:
```ts
 //defaults
pathtracer.denoiseAt = 6;
pathtracer.generateAux= true;
// later
pathtracer.renderSample();
// optional
pathtracer.denoiseSample(); // denoises current sample
pathtracer.setAux(albedoTexture, normalTexture);
```

**Comment by gkjohnson** [comment]
> Do you want to assume the pathtracer will be generating the albedos and normals by default or leave that as an external object the user has to setup/use?

I think the path tracer should handle all of this internally to make it as simple as possible to use. I think the API can be as simple as this:

```js
pathtracer.maxSamples = 10 // defaults to 20 or so?
pathtracer.enableDenoiser = true // defaults to true?
```

The path tracer will stop rendering once `maxSamples` is reached regardless of whether the denoiser is enabled. And if the denoiser _is_ enabled then denoising will occur once the cap is hit.

Setting `enableDenoiser` to false and `maxSamples` to Infinity gets the current behavior of the pathtracer. I'm torn on whether the denoiser should default to true or not. But that can be decided later.

**Comment by andygoodluck** [comment]
Hi @gkjohnson,

I would like to work on implementing a denoiser for three-gpu-pathtracer.

**Background:**
- Experienced with WebGL/Three.js rendering
- Familiar with image processing and denoising algorithms
- Knowledge of SVGF and OIDN denoising techniques
- GitHub: @andygoodluck

**Technical Approach:**
1. Research glslSmartDeNoise implementation
2. Integrate Blackman-Harris pixel filter
3. Implement temporal accumulation for noise reduction
4. Add configurable denoise parameters
5. Optimize for real-time performance

**References:**
- glslSmartDeNoise (BrutPitt)
- SVGF denoising paper
- OIDN integration patterns

**Timeline:** 2-3 weeks

**Deliverables:**
- Integrated denoiser with configurable settings
- Performance benchmarks
- Documentation and examples
- Unit tests

Ready to start immediately!
